windows: Add machine.UART support for serial COM ports.

kimstik opened the pull request:
### Summary

Implement machine.UART class for Windows port, providing access to serial COM ports with an API compatible with other MicroPython ports.

Features:
- Support for COM1-COM255 via integer or string identifier
- Standard UART parameters: baudrate, bits, parity, stop, flow control
- Configurable timeouts (timeout, timeout_char)
- Configurable buffer sizes (rxbuf, txbuf)
- Stream protocol support (read, write, poll, flush)
- UART.list() class method to enumerate available COM ports
- sendbreak() and readchar()/writechar() support

Implementation uses Win32 API (CreateFile, DCB, COMMTIMEOUTS) with the MICROPY_PY_MACHINE_UART_INCLUDEFILE pattern for integration.

### Testing

  **CI Testing:**
  - GitHub Actions: All builds pass (MSVC 2017/2019/2022 x86/x64, MinGW32/64)
  - Test suite: 823/824 tests pass (see skip note below)

  **Hardware Testing:**
  - Tested with physical USB-to-serial adapter

  **Skipped Tests:**
  - `extmod/machine_uart_tx.py` - requires physical UART with loopback wiring
  - `extmod/machine_uart_irq_txidle.py` - requires IRQ support (not implemented)

  Note: `mpremote` build may fail in CI due to upstream issue with
  `hatch-vcs` version detection in forked repository (not related
  to this PR).

```
from machine import UART
import sys

print("Ports:", ports := UART.list())

if ports:
    uart = UART(ports[0], 115200, timeout=100)
    print(uart)
    uart.write(b"Hello from MicroPython\r\n")
    while True:
        d = uart.read(256)
        if d:
            sys.stdout.write(d)

# TODO:
# * select.poll(stdin)
# * sys.stdin.any()

```

Commit:
windows: Add machine.UART support for serial COM ports.

Implement machine.UART class for Windows port, providing access to
serial COM ports with an API compatible with other MicroPython ports.

Features:
- Support for COM1-COM255 via integer or string identifier
- Standard UART parameters: baudrate, bits, parity, stop, flow control
- Configurable timeouts (timeout, timeout_char)
- Configurable buffer sizes (rxbuf, txbuf)
- Stream protocol support (read, write, poll, flush)
- UART.list() class method to enumerate available COM ports
- sendbreak() and readchar()/writechar() support

Implementation uses Win32 API (CreateFile, DCB, COMMTIMEOUTS) with
the MICROPY_PY_MACHINE_UART_INCLUDEFILE pattern for integration.

Signed-off-by: Konstantin Kim <kimstik@gmail.com>

stinos commented on ports/windows/machine_uart.c:
```diff
@@ -0,0 +1,510 @@
+#define DEFAULT_UART_TIMEOUT (0)
+#define DEFAULT_UART_TIMEOUT_CHAR (0)
+
+#define UART_HWCONTROL_RTS (1)
```
Windows provides definitions like `RTS_CONTROL_HANDSHAKE` so seems better to use those here.

kimstik replied:
The current approach uses bit flags for Python API so users can do flow=UART.RTS|UART.CTS

stinos commented on ports/windows/machine_uart.c:
```diff
@@ -0,0 +1,510 @@
+    if (args[ARG_baudrate].u_int > 0) {
+        self->baudrate = args[ARG_baudrate].u_int;
+    }
+    if (args[ARG_bits].u_int >= 5 && args[ARG_bits].u_int <= 8) {
```
The same logic already exists in a function above, can that not be reused?

kimstik replied:
this is the standard MicroPython pattern
make_new sets defaults, then calls init_helper to apply user arguments
No actual duplication here: make_new handles object creation + port name parsing, init_helper handles parameter parsing.
Unless you mean something else?

kimstik commented on tests/run-tests.py:
```diff
@@ -210,6 +210,11 @@ def open(self, path, mode):
         "thread/thread_shared2.py",
     ),
+    "win32": (
+        # These tests require physical UART hardware with loopback wiring.
```
This is more complex than it appears.
The target_wiring mechanism was designed for remote targets (pyboard interface) where scripts are injected into sys.modules. For local runs (unix/windows), detect_target_wiring_script() is skipped entirely, so from target_wiring import ... fails with ImportError.

Supporting local target wiring requires changes to run-tests.py infrastructure. Also unclear if CI has COM port access - tests will fail at UART instantiation anyway if not.

Keeping UART tests in skip list until test infrastructure is updated.

Commit:
windows/machine_uart: Address review feedback from @stinos.

- Fix UART.list() to check ERROR_NO_MORE_ITEMS specifically and skip
  entries that fail due to buffer too small instead of breaking early.
- Make sendbreak duration baudrate-dependent (13 bit periods) like rp2.
- Disable DTR by default to avoid resetting Arduino/ESP devices.
- Add UART_PORT_NAME_MAX constant for consistent buffer sizes.
- Remove DEFAULT_UART_* macros, use inline values in one place.
- Add comments for flow control flags and buffer sizes.

Signed-off-by: Konstantin Kim <kimstik@gmail.com>

kimstik committed:
tests/run-tests.py: Update comment for UART hardware requirement.

no loopback requirement

Signed-off-by: Konstantin Kim <kimstik@gmail.com>

Josverl commented:
Should there be an option to filter out the Bluetooth serial.?

Bluetooth serial: when Windows creates outgoing/incoming RFCOMM ports, they appear as COMx and are listed in the registry.

AFAIK the registry does not provide sufficient detail for that.

Commit:
windows/machine_uart: Use portNameSize instead of strlen in UART.list().

RegEnumValueA already returns the string length in portNameSize,
so strlen() call is redundant.

Signed-off-by: Konstantin Kim <kimstik@gmail.com>

stinos commented on ports/windows/machine_uart.c:
```diff
@@ -0,0 +1,510 @@
+    dcb.fRtsControl = (self->flow & UART_HWCONTROL_RTS) ? RTS_CONTROL_HANDSHAKE : RTS_CONTROL_ENABLE;
+    dcb.fOutX = FALSE;
+    dcb.fInX = FALSE;
+    dcb.fDtrControl = DTR_CONTROL_ENABLE;
```
So should this be changed? I'm really not sure. I've had issues with this in the past (i.e. unwanted restarts of the Arduino) so maybe this should even be configurable?

kimstik commented:
> > Feel free to let me know if I missed anything.
> 
> The code looks ok but to be sure it should be tested with a couple of devices at least I guess and I don't have anything ready atm. I assume you tested this with actual hardware: what did you use?

Yes, I did some light testing with what I have (usb-com dongles, com2com, ...) and within the time frame I could spare. Extra testing never hurts, so any additional testing will be valuable.
The original idea was purely practical - to create a base for RS485/Modbus.

Commit:
windows/machine_uart: Ensure null-termination in UART.list().

Per MS documentation, registry strings may not be properly
null-terminated. Write null at portNameSize position to handle
both terminated and unterminated cases correctly.

Signed-off-by: Konstantin Kim <kimstik@gmail.com>

kimstik commented:
Did I miss something?
